### ISSUE TITLE
Generic parameter type checking stops too early

### ISSUE THREAD

**Issue by devongovett** [microsoft/TypeScript#59049]
### 🔎 Search Terms

generic, early, permature, defer

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ssl=10&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgEoRgeQEYCsIJgA8AKgHzIDeAUMncggK5RQTgBcyJyAPsiIwA2g6gF9q1ACYFBcVshiMQhYAHsQySAGcwACgAendFjwFiAgLbZoZAJScAbquCSA3FJlyUi5WDUaAcwgwY117NAwcfEIiS2soXn4hQTJ3akFg5AdkAF5kIJCMMPdtPQdbdyA

### 💻 Code

```ts
interface RefObject<T> {
    current: T | null
}

declare function test(x: RefObject<number>): void;
declare function getRef(): RefObject<number | null>;

let v = getRef();
test(v);
```

### 🙁 Actual behavior

In the above code, `RefObject` is defined to always accept `null` in its `current` property. However, if one explicitly writes `RefObject<number | null>` and passes the result to a parameter accepting `RefObject<number>` TS reports a type error even though the types of the `current` property would have been the same. This happens with any type, not just null (e.g. if the definition of current was `T | string`.

### 🙂 Expected behavior

Since the `current` properties are equivalent after substituting the generic parameters, there should be no error. It seems like TS is stopping as soon as it sees `RefObject<number | null>` vs `RefObject<number>` even though substituting the generic parameter would result in the same type. This can be confirmed by creating a copy of the RefObject type like this:

```ts
interface RefObject<T> {
    current: T | null
}

interface RefObject2<T> {
    current: T | null
}

declare function test(x: RefObject<number>): void;
declare function getRef(): RefObject2<number | null>;

let v = getRef();
test(v);
```

[Playground](https://www.typescriptlang.org/play/?ts=5.5.2&ssl=14&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgEoRgeQEYCsIJgA8AKgHzIDeAUMncggK5RQTgBcyJyAPsiIwA2g6gF9q1UJFiIU6LHgJgATKQo16DZqw5de-ISPHUAJgUFxWyGIxCFgAexDJIAZzAAKAB6d5OfIREAgC22NBkAJScAG4OwCYA3Kbmlig2dmCOzgDmEGDyHlFoGP5KqiFhUPoCwmRJ1IJ5yNHIALzIufkYhUluntERSUA)

### Additional information about the issue

This affects the change in React's types between version 18 and 19, where the `RefObject` type changed `current` from `T | null` to only `T`. The uses of `RefObject` should add `null` instead, e.g. `RefObject<HTMLDivElement | null>`. Making libraries compatible with both the React 18 and React 19 types is difficult due to this TS issue. If a library updates its implementation to React 19, including `RefObject<HTMLElement | null>` in its definitions, consumers still using React 18 types will see TS errors.

**Comment by RyanCavanaugh** [comment]
This is tricky. By default this is done with a variance-based short circuiting; *not* doing this would be catastrophic to performance. Since it's only needed to probe the type deeper if it's the case that `T` is *only* referenced in places where it's unioned with `null`, a better workaround would be to do something like this
```ts
type RefObject<T> = _RefObject<Exclude<T, null>>;
interface _RefObject<T> {
    current: T | null
}
```

**Comment by devongovett** [comment]
Ah interesting. I figured it was for performance. I guess your workaround would need to be added to the _old_ versions of the React types (e.g. 18, 17, 16, etc), since React 19 removed the `| null`. Not sure if they'd be open to that and it would require everyone update to a new patch.

The other alternative we were considering for our library was to have our own copy of `RefObject` locally and use that everywhere instead of the one imported from `react`. That should also work but it means that each library that wants to maintain compatibility with multiple React versions will need to do that.

**Comment by eps1lon** [comment]
> Not sure if they'd be open to that and it would require everyone update to a new patch.

If it's not breaking, you can file a PR and we see if it integrates cleanly.

```tsx
type RefObject<T> = _RefObject<Exclude<T, null>>;
interface _RefObject<T> {
  current: T | null
}
```

This is just really gnarly and given how unpredictable type emission is in TypeScript is with numerous issues closed as intended or left unaddressed, I'd want to consider first if we want to merge this. There are already enough types to deal with as learners and adding some weird `_` prefixed types just makes it worse if they leak into user code.

**Comment by eps1lon** [comment]
> not ideal that projects would need to upgrade the 16/17/18 types in order to fix compatibility with libraries built for the 19 types either.

I don't think this is an issue. I think if we can't even recommend updating libraries by a patch version, we've lost the ability to move the ecosystem forward. We can recommend upgrading React types by a patch once, or teach each library how to add a compat layer and then rely on consumers to upgrade each of these libraries? Seems way more simple to fix this in React types.

But if it requires this workaround that may leak into emitted types, it doesn't seem like a good workaround. Regardless of whether we do this in React types or 3rd party libraries. It would be more helpful if TypeScript would either fix this or improve type emission so that libraries are in control of what types are public and private.

Why can't you upgrade all your callsites of `RefObject` to use `RefObject<T | null>` right now ignoring anything that comes in React 19? You said it causes errors but I'm not seeing any in the playground nor do I see an explainer why it should error. The sample code you posted could be fixed if all `RefObject` usages would use `RefObject<T | null>`.

**Comment by devongovett** [comment]
> Why can't you upgrade all your callsites of RefObject to use RefObject<T | null> right now

Yeah that's what we've done. The problem occurs when passing a ref typed as `RefObject<T | null>` to anything that accepts `React.RefAttributes<T>` in previous versions of React, including builtin JSX elements like `<div>`, and any `forwardRef` components.

Here's a more practical example of what I mean: [Playground](https://www.typescriptlang.org/play/?ts=5.5.2#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDeyBmA8gEYBWSGAvnPlBCHAORSoaMDcAUJwCYUA2KFjQCuAOwzAIYuCIDOSABIQIAawAqACxQxkMEVDFyCACgCUALkRIiZCjAA8i9QFkAMgCERMGNICi-EggSGLwAD5wYiL8-AB8XJz44pLScOpIcjDmuJxweXCB8Cz4cAC8sgrKalo6egZGpmZc+XAs+oZwDjzAAG6tNqU4xdQA9LGclJxAA)

In React 19, `ForwardedRef` and `RefAttributes` include `RefObject<T | null>` but in previous versions of React, the `| null` is _inside_ the definition of `RefObject`. So due to this TS limitation, if we add `T | null` to our definition, we are compatible with React 19 but not 18 or previous, and if we don't add it we are compatible with previous but not 19.

**Comment by eps1lon** [comment]
That is a problem that I would love being able to fix in React types. But this `_RefObject` types appears in emitted declarations and we just created another problem that'll take years to get rid of when TypeScript changes behavior in that area. And there's the cost of having to learn/teach why we have this intermediate type.

**Comment by devongovett** [comment]
Yeah I agree. Ideally TS would fix this issue and prioritize correctness over performance. But in the meantime we'll need a recommendation because I would imagine we are not the only library that will be affected by this.

Another option I thought of would be to add a new `NullableRefObject` type in both old and new versions of the React types. In old react it could be an alias:

```ts
// In old react types
type NullableRefObject<T> = RefObject<T>;
```

and in React 19 it could add `null` to T:

```ts
In React 19
type NullableRefObject<T> = RefObject<T | null>;
```

Then libraries could use `NullableRefObject` everywhere. Perhaps slightly less hacky but still requires updating many versions of the types, and teaching a new type. 🤔

**Comment by RyanCavanaugh** [comment]
I'm not *endorsing* this as a great solution, but it works to get structural comparison as a fallback (thus preserving the perf benefits of variance measurement)
```ts
declare const UnreliableSymbol: unique symbol;
type Struct<T> = `${T & string}`; 

type DataOrNull< /* unreliable out */ T> = {
  data: T | null;
  [UnreliableSymbol]: Struct<T>;
}

declare function check(s: DataOrNull<string>): void;

declare let foo: DataOrNull<string | null>;
check(foo); // ok
```

**Comment by CraigMacomber** [comment]
I believe I hit a roughly equivalent issue. Since I produces a minimal repro for it and its a bit different than the one above I figured I might as well include it here incase its useful:

[Playground](https://www.typescriptlang.org/play/?ssl=24&ssc=1&pln=1&pc=1#code/C4TwDgpgBAKg9gawgOygXigRgNxQPR5QDKcAttAM6QDGAlgIYA2UAbkwK4QBQXBUAglFCQoAd1rAAFlGpkwjWtXrAIAE1b0ATg2TUIAOl6EAsnArAoAcxQRNTGfQrRaFMZJRQABvCTJPUFygmRjhRNUNhaABhTQhlCAAeGAA+dCgAbygADwAuWEQPCCyVZFVXGCgAfih2ZARkUNQ82CgAX2wePhhROADVFGBFe1pkFU0AM3o9CkM+ACF2Sys4CFd6UXoQAPGhdycg2KFwaH7xkYlaOGRXEfM49TgdkbHJ6cNn21foAA1MJNSiiUylAYnEVP8Mq0uB8JlMfgAmCGAlDA0HxCHpKFcSJQACSyDA7AsGHhHT4UTkROgwB6fQGQ2YkVc7AoIyWvyCpSg33h2OOUAASqt2IxgJg0r8EvjCcBkh0cUKKCLgPCJYjpUS5Z1CDB3HTRgyjpA1ocGhYIABHdhMAA0UAamgO1L1tH6BqUzBhXxmfJEAH1qO5qAhMKqMLErbRYvwKKzLMh6AAjRgQeAJRXKzB2nlJArIZJanEBoMh8Xhy3sKMQGNxhPJ1NwdPC0VZ7l-HwoAu4PgZ0VQcaaMhtgKuM1BWO0eNJlNCXq-O1XRhbHm+6DFiDB+FhqARyvRidT+tp3sq7OIjv5uX4Qgn-uD0jc1WBMeOWvT6lz+EL5BLturqDrpuZY7hWVY1pOdYpsezanm2ua+F2PA4ruYEHpBDYJH6gjIqUrhzHacypBgwCaJwHRAA)

```typescript
type Token = 1; // Some special value

// A type with complicated variance.
// Most general case is when `Token` is allowed.
type Create<T> = { x: Token extends T ? unknown :  T };

// Two identical interfaces.
// Bug goes away if these are type definitions instead of interfaces.
interface X1<T> extends Create<T> {}
interface X2<T> extends Create<T> {}

type Input = 2;
// Compute two identical types using X1 and X2
type Result1 = X1<Input>;
type Result2 = X2<Input>;

// The identical types are not equal, nor are the identical interfaces.
type _check12 = requireAssignableTo<Result1, X2<Token>>;
type _check11 = requireAssignableTo<Result1, X1<Token>>; // Result from X1 is not assignable to X1, only X2
type _check22 = requireAssignableTo<Result2, X2<Token>>; // Result from X2 is not assignable to X2, only X1
type _check21 = requireAssignableTo<Result2, X1<Token>>;

type requireAssignableTo<_A extends B, B> = true;
```
